File: MiniDOMWebCtrls/DMD/Forms/WebControlStyle.cs
using System;
using DMD;
using DMD.XML;

namespace minidom.Forms
{
    [Serializable]
    public class WebControlStyle : IDMDXMLSerializable
    {
        [NonSerialized]
        private WebControl m_Owner;
        public string Position;
        private string m_Left;
        private string m_Top;
        private string m_Bottom;
        private string m_Right;
        private string m_Width;
        private string m_Height;
        public WebControlBackGround Background;
        public WebControlBorder Border;
        public string Color;
        public string TextAlign;
        public string zIndex;
        public string Visibility;
        public string Overflow;
        public string WhiteSpace;
        public string MinWidth;
        public string MaxWidth;
        public string MinHeight;
        public string MaxHeight;
        public string Display;
        public string VerticalAlign = "";

        public WebControlStyle()
        {
            DMDObject.IncreaseCounter(this);
            Background = new WebControlBackGround(this);
            Border = new WebControlBorder(this);
        }

        public WebControlStyle(WebControl owner) : this()
        {
            m_Owner = owner;
        }

        public string Left
        {
            get
            {
                return m_Left;
            }

            set
            {
                m_Left = NormalizeSize(value);
            }
        }

        public string Top
        {
            get
            {
                return m_Top;
            }

            set
            {
                m_Top = NormalizeSize(value);
            }
        }

        public string Right
        {
            get
            {
                return m_Right;
            }

            set
            {
                m_Right = NormalizeSize(value);
            }
        }

        public string Bottom
        {
            get
            {
                return m_Bottom;
            }

            set
            {
                m_Bottom = NormalizeSize(value);
            }
        }

        public string Width
        {
            get
            {
                return m_Width;
            }

            set
            {
                m_Width = NormalizeSize(value);
            }
        }

        public string Height
        {
            get
            {
                return m_Height;
            }

            set
            {
                m_Height = NormalizeSize(value);
            }
        }

        private string NormalizeSize(string value)
        {
            value = Strings.Trim(value);
            if (string.IsNullOrEmpty(value))
                return "";
            string ch = DMD.Strings.Right(value, 1);
            if (DMD.Strings.Compare(ch, "0", false) >= 0 && DMD.Strings.Compare(ch, "9", false) <= 0)
                value = value + "px";
            return value;
        }

        public override string ToString()
        {
            string ret = DMD.Strings.vbNullString;
            string tmp;
            if (!string.IsNullOrEmpty(TextAlign))
                ret = DMD.Strings.Combine(ret, "text-align:" + TextAlign, ";");
            if (!string.IsNullOrEmpty(Position))
                ret = DMD.Strings.Combine(ret, "position:" + Position, ";");
            if (!string.IsNullOrEmpty(Left))
                ret = DMD.Strings.Combine(ret, "left:" + Left, ";");
            if (!string.IsNullOrEmpty(Top))
                ret = DMD.Strings.Combine(ret, "top:" + Top, ";");
            if (!string.IsNullOrEmpty(Right))
                ret = DMD.Strings.Combine(ret, "right:" + Right, ";");
            if (!string.IsNullOrEmpty(Bottom))
                ret = DMD.Strings.Combine(ret, "bottom:" + Bottom, ";");
            if (!string.IsNullOrEmpty(Width))
                ret = DMD.Strings.Combine(ret, "width:" + Width, ";");
            if (!string.IsNullOrEmpty(Height))
                ret = DMD.Strings.Combine(ret, "height:" + Height, ";");
            if (!string.IsNullOrEmpty(Overflow))
                ret = DMD.Strings.Combine(ret, "overflow:" + Overflow, ";");
            if (!string.IsNullOrEmpty(WhiteSpace))
                ret = DMD.Strings.Combine(ret, "white-space:" + WhiteSpace, ";");
            if (!string.IsNullOrEmpty(MinWidth))
                ret = DMD.Strings.Combine(ret, "min-width:" + MinWidth, ";");
            if (!string.IsNullOrEmpty(MaxWidth))
                ret = DMD.Strings.Combine(ret, "max-width:" + MaxWidth, ";");
            if (!string.IsNullOrEmpty(MinHeight))
                ret = DMD.Strings.Combine(ret, "min-height:" + MinHeight, ";");
            if (!string.IsNullOrEmpty(MaxHeight))
                ret = DMD.Strings.Combine(ret, "max-height:" + MaxHeight, ";");
            if (!string.IsNullOrEmpty(Display))
                ret = DMD.Strings.Combine(ret, "display:" + Display, ";");
            if (!string.IsNullOrEmpty(VerticalAlign))
                ret = DMD.Strings.Combine(ret, "vertical-align:" + VerticalAlign, ";");
            tmp = Border.ToString();
            if (!string.IsNullOrEmpty(tmp))
                ret = DMD.Strings.Combine(ret, "border:" + tmp, ";");
            tmp = Background.ToString();
            if (!string.IsNullOrEmpty(tmp))
                ret = DMD.Strings.Combine(ret, "background:" + tmp, ";");
            if (!string.IsNullOrEmpty(Color))
                ret = DMD.Strings.Combine(ret, "color:" + Color, ";");
            if (!string.IsNullOrEmpty(zIndex))
                ret = DMD.Strings.Combine(ret, "z-index:" + zIndex, ";");
            if (!string.IsNullOrEmpty(Visibility))
                ret = DMD.Strings.Combine(ret, "visibility:" + Visibility, ";");
            return ret;
        }

        public void SetFieldInternal(string fieldName, object fieldValue)
        {
            switch (fieldName ?? "")
            {
                case "Position":
                    {
                        Position = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Left":
                    {
                        Left = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Top":
                    {
                        Top = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Bottom":
                    {
                        Bottom = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Right":
                    {
                        Right = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Width":
                    {
                        Width = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Height":
                    {
                        Height = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Color":
                    {
                        Color = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "TextAlign":
                    {
                        TextAlign = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "zIndex":
                    {
                        zIndex = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Visibility":
                    {
                        Visibility = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Overflow":
                    {
                        Overflow = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "WhiteSpace":
                    {
                        WhiteSpace = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "MinWidth":
                    {
                        MinWidth = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "MaxWidth":
                    {
                        MaxWidth = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "MinHeight":
                    {
                        MinHeight = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "MaxHeight":
                    {
                        MaxHeight = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Display":
                    {
                        Display = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "VerticalAlign":
                    {
                        VerticalAlign = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Background":
                    {
                        Background = (WebControlBackGround)fieldValue;
                        break;
                    }

                case "Border":
                    {
                        Border = (WebControlBorder)fieldValue;
                        break;
                    }
            }
        }

        public void XMLSerialize(XMLWriter writer)
        {
            writer.WriteAttribute("Position", Position);
            writer.WriteAttribute("Left", Left);
            writer.WriteAttribute("Top", Top);
            writer.WriteAttribute("Bottom", Bottom);
            writer.WriteAttribute("Right", Right);
            writer.WriteAttribute("Width", Width);
            writer.WriteAttribute("Height", Height);
            writer.WriteAttribute("Color", Color);
            writer.WriteAttribute("TextAlign", TextAlign);
            writer.WriteAttribute("zIndex", zIndex);
            writer.WriteAttribute("Visibility", Visibility);
            writer.WriteAttribute("Overflow", Overflow);
            writer.WriteAttribute("WhiteSpace", WhiteSpace);
            writer.WriteAttribute("MinWidth", MinWidth);
            writer.WriteAttribute("MaxWidth", MaxWidth);
            writer.WriteAttribute("MinHeight", MinHeight);
            writer.WriteAttribute("MaxHeight", MaxHeight);
            writer.WriteAttribute("Display", Display);
            writer.WriteAttribute("VerticalAlign", VerticalAlign);
            writer.WriteTag("Background", Background);
            writer.WriteTag("Border", Border);
        }

        protected internal virtual void SetOwner(WebControl owner)
        {
            m_Owner = owner;
        }

        ~WebControlStyle()
        {
            DMDObject.DecreaseCounter(this);
        }
    }
}

File: MiniDOMWebCtrls/DMD/Forms/WebControlStyle.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.WebSite
Imports minidom.Databases


Namespace Forms

    <Serializable> _
    Public Class WebControlStyle
        Implements XML.IDMDXMLSerializable

        <NonSerialized> Private m_Owner As WebControl
        Public Position As String
        Private m_Left As String
        Private m_Top As String
        Private m_Bottom As String
        Private m_Right As String
        Private m_Width As String
        Private m_Height As String
        Public Background As WebControlBackground
        Public Border As WebControlBorder
        Public Color As String
        Public TextAlign As String
        Public zIndex As String
        Public Visibility As String
        Public Overflow As String
        Public WhiteSpace As String
        Public MinWidth As String
        Public MaxWidth As String
        Public MinHeight As String
        Public MaxHeight As String
        Public Display As String
        Public VerticalAlign As String = ""

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.Background = New WebControlBackGround(Me)
            Me.Border = New WebControlBorder(Me)
        End Sub

        Public Sub New(ByVal owner As WebControl)
            Me.New()
            Me.m_Owner = owner
        End Sub

        Public Property Left As String
            Get
                Return Me.m_Left
            End Get
            Set(value As String)
                Me.m_Left = NormalizeSize(value)
            End Set
        End Property

        Public Property Top As String
            Get
                Return Me.m_Top
            End Get
            Set(value As String)
                Me.m_Top = NormalizeSize(value)
            End Set
        End Property

        Public Property Right As String
            Get
                Return Me.m_Right
            End Get
            Set(value As String)
                Me.m_Right = NormalizeSize(value)
            End Set
        End Property

        Public Property Bottom As String
            Get
                Return Me.m_Bottom
            End Get
            Set(value As String)
                Me.m_Bottom = NormalizeSize(value)
            End Set
        End Property

        Public Property Width As String
            Get
                Return Me.m_Width
            End Get
            Set(value As String)
                Me.m_Width = NormalizeSize(value)
            End Set
        End Property

        Public Property Height As String
            Get
                Return Me.m_Height
            End Get
            Set(value As String)
                Me.m_Height = NormalizeSize(value)
            End Set
        End Property

        Private Function NormalizeSize(ByVal value As String) As String
            value = Trim(value)
            If (value = "") Then Return ""
            Dim ch As String = Strings.Right(value, 1)
            If (ch >= "0" AndAlso ch <= "9") Then value = value & "px"
            Return value
        End Function

        Public Overrides Function ToString() As String
            Dim ret As String = vbNullString
            Dim tmp As String
            If (Me.TextAlign <> "") Then ret = Strings.Combine(ret, "text-align:" & Me.TextAlign, ";")
            If (Me.Position <> "") Then ret = Strings.Combine(ret, "position:" & Me.Position, ";")
            If (Me.Left <> "") Then ret = Strings.Combine(ret, "left:" & Me.Left, ";")
            If (Me.Top <> "") Then ret = Strings.Combine(ret, "top:" & Me.Top, ";")
            If (Me.Right <> "") Then ret = Strings.Combine(ret, "right:" & Me.Right, ";")
            If (Me.Bottom <> "") Then ret = Strings.Combine(ret, "bottom:" & Me.Bottom, ";")
            If (Me.Width <> "") Then ret = Strings.Combine(ret, "width:" & Me.Width, ";")
            If (Me.Height <> "") Then ret = Strings.Combine(ret, "height:" & Me.Height, ";")
            If (Me.Overflow <> "") Then ret = Strings.Combine(ret, "overflow:" & Me.Overflow, ";")
            If (Me.WhiteSpace <> "") Then ret = Strings.Combine(ret, "white-space:" & Me.WhiteSpace, ";")
            If (Me.MinWidth <> "") Then ret = Strings.Combine(ret, "min-width:" & Me.MinWidth, ";")
            If (Me.MaxWidth <> "") Then ret = Strings.Combine(ret, "max-width:" & Me.MaxWidth, ";")
            If (Me.MinHeight <> "") Then ret = Strings.Combine(ret, "min-height:" & Me.MinHeight, ";")
            If (Me.MaxHeight <> "") Then ret = Strings.Combine(ret, "max-height:" & Me.MaxHeight, ";")
            If (Me.Display <> "") Then ret = Strings.Combine(ret, "display:" & Me.Display, ";")
            If (Me.VerticalAlign <> "") Then ret = Strings.Combine(ret, "vertical-align:" & Me.VerticalAlign, ";")

            tmp = Me.Border.ToString
            If tmp <> vbNullString Then ret = Strings.Combine(ret, "border:" & tmp, ";")
            tmp = Me.Background.ToString
            If (tmp <> vbNullString) Then ret = Strings.Combine(ret, "background:" & tmp, ";")
            If (Me.Color <> vbNullString) Then ret = Strings.Combine(ret, "color:" & Me.Color, ";")
            If (Me.zIndex <> vbNullString) Then ret = Strings.Combine(ret, "z-index:" & Me.zIndex, ";")
            If (Me.Visibility <> vbNullString) Then ret = Strings.Combine(ret, "visibility:" & Me.Visibility, ";")
            Return ret
        End Function

        Public Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Position" : Me.Position = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Left" : Me.Left = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Top" : Me.Top = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Bottom" : Me.Bottom = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Right" : Me.Right = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Width" : Me.Width = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Height" : Me.Height = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Color" : Me.Color = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "TextAlign" : Me.TextAlign = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "zIndex" : Me.zIndex = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Visibility" : Me.Visibility = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Overflow" : Me.Overflow = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "WhiteSpace" : Me.WhiteSpace = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MinWidth" : Me.MinWidth = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MaxWidth" : Me.MaxWidth = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MinHeight" : Me.MinHeight = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "MaxHeight" : Me.MaxHeight = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Display" : Me.Display = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "VerticalAlign" : Me.VerticalAlign = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Background" : Me.Background = fieldValue
                Case "Border" : Me.Border = fieldValue
            End Select
        End Sub

        Public Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Position", Me.Position)
            writer.WriteAttribute("Left", Me.Left)
            writer.WriteAttribute("Top", Me.Top)
            writer.WriteAttribute("Bottom", Me.Bottom)
            writer.WriteAttribute("Right", Me.Right)
            writer.WriteAttribute("Width", Me.Width)
            writer.WriteAttribute("Height", Me.Height)
            writer.WriteAttribute("Color", Me.Color)
            writer.WriteAttribute("TextAlign", Me.TextAlign)
            writer.WriteAttribute("zIndex", Me.zIndex)
            writer.WriteAttribute("Visibility", Me.Visibility)
            writer.WriteAttribute("Overflow", Me.Overflow)
            writer.WriteAttribute("WhiteSpace", Me.WhiteSpace)
            writer.WriteAttribute("MinWidth", Me.MinWidth)
            writer.WriteAttribute("MaxWidth", Me.MaxWidth)
            writer.WriteAttribute("MinHeight", Me.MinHeight)
            writer.WriteAttribute("MaxHeight", Me.MaxHeight)
            writer.WriteAttribute("Display", Me.Display)
            writer.WriteAttribute("VerticalAlign", Me.VerticalAlign)
            writer.WriteTag("Background", Me.Background)
            writer.WriteTag("Border", Me.Border)
        End Sub

        Protected Friend Overridable Sub SetOwner(ByVal owner As WebControl)
            Me.m_Owner = owner
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Namespace